File: play_game.jl
using Plots
using Dates
using DelimitedFiles

starttime = Dates.now()

max_tile = 9
one_dice = false
one_dice_num = 6    #when open tiles are equal to or less than this number, one dice is used

reduce_tile_drops = true
tiles_spaced = true

dice_num = 6

function roll_dice(arr)
    if one_dice & (maximum(arr) <= one_dice_num)
        dice = rand(1:dice_num)
    else
        dice = rand(1:dice_num) + rand(1:dice_num)
    end
    return dice
end

function possible_drop(arr, drop_num)
    in_arr = issubset(drop_num, arr)
    return in_arr
end

function drop_tiles(arr, drop_num)
    arr_len = length(arr)
    for i in 1:arr_len
        arr_i = arr[i]
        if arr_i in drop_num
            arr[i] = 0
        end
    end
    return arr
end

function combo_gen(numbers, target, partial = Int16[], partial_sum = 0)

    if partial_sum == target
        global res = vcat(res, [partial])
        return res
    end
    if partial_sum >= target
        return nothing
    end

    num_len = length(numbers)
    if reduce_tile_drops
        lst = enumerate(numbers)
    else
        lst = enumerate(reverse(numbers))
    end

    for i in lst
        if reduce_tile_drops
            remaining = numbers[i[1]+1:num_len]
        else
            remaining = numbers[1:(num_len - i[1])]
        end

        combo_gen(remaining, target, vcat(partial, i[2]), partial_sum + i[2])
    end

end


function possible_combos(numbers, target)
    global res = Int16[]
    combo_gen(numbers, target)
    return res
end

function reverse_combos(arr)
    arr_len = [length(x) for x in arr]
    println(arr)
    println(arr_len[arr_len == 2])
end

function make_combos(arr)
    output = Int16[]
    for dice in 6:7
        output = vcat(output, [possible_combos(arr, dice)])
    end
    return output
end

x = make_combos(Int16[x for x in 1:max_tile])
print(reverse_combos(x[1]))


function game_play(arr, combos)
    game_going = true
    round_score = zeros(Int16, max_tile + 1)
    round_score[1] = sum(arr)

    idx = 1
    while (sum(arr) > 0) & game_going
        dice = roll_dice(arr)
        moves = combos[dice]

        combo_looking = false
        n = 0
        while (~combo_looking) & (n < length(moves))
            n += 1
            drops = moves[n]
            if possible_drop(arr, drops)
                idx += 1
                arr = drop_tiles(arr, drops)
                round_score[idx] = sum(arr)
                combo_looking = true
            end
        end
        game_going = combo_looking
    end
    final_score = round_score[idx]
    idx += 1
    for i in idx:max_tile
        round_score[i] = final_score
    end
    return round_score, arr
end


# println("Prep")
# new_arr = Int16[x for x in 1:max_tile]
# combinations = make_combos(new_arr)
# println(combinations)

function mult_games(n)
    println("Running games")
    scores = zeros(Int16, (n, max_tile+1))
    results = zeros(Int16, (n, max_tile))
    for i in 1:n
        arr = copy(new_arr)
        val, arr = game_play(copy(new_arr), combinations)
        scores[i, :] = val
        results[i, :] = arr
    end
    return scores, results
end

# scores, results = mult_games(1000)

# println("Prepping Results")
# x_val = [x for x in 0:sum(1:max_tile)]
# y = copy(x_val)

# for i in x_val
#     y[i+1] = length([x for x in scores if x == i])
# end 

# y = y./sum(y)

# println(string(y[1] * 100) * "%")
# print(Dates.now() - starttime)

# loc = "/Users/ushhamilton/Documents/03 Programming/Julia/Shut The Box/Outputs/"
# writedlm(loc * "Score_Output.csv", scores, ',')
# writedlm(loc * "Array_Output.csv", results, ',')

File: play_game.py
import numpy as np
import random
import matplotlib.pyplot as plt
import datetime as dt


class play_game:
    # Game parameters
    max_tile = 9
    one_dice = False
    one_dice_num = 6    #Number under which one dice can be used

    dice_num = 6    # Size of dice 1-x
    def __init__(self):
        self.arr = [x for x in range(1, self.max_tile + 1)]

    def roll_dice(self):
        if self.one_dice and (max(self.arr) <= self.one_dice_num):
            dice = random.randint(1, self.one_dice_num)
        else:
            dice = random.randint(1, self.one_dice_num) + random.randint(1, self.one_dice_num)
        
        return dice

    def possible_drop(self, drop_nums):
        #Check incrementing correctly
        increment = True
        for i in range(len(drop_nums)-1):
            increment = increment and (drop_nums[i] > drop_nums[i + 1])
        #Check if the given array is permitted for dropping tiles
        in_arr = True
        for t in drop_nums:
            if t not in self.arr:
                in_arr = False
        return in_arr and increment
    
    def drop_tiles(self, drop_nums):
        #Drop given tile numbers
        self.arr = [0 if x in drop_nums else x for x in self.arr]
        return sum(self.arr)

    def possible_combos(self, numbers, target, partial=[], partial_sum=0):
        numbers = [x for x in numbers if x > 0]
        if partial_sum == target:
            yield partial
        if partial_sum >= target:
            return
        for i, n in reversed(list(enumerate(numbers))):
            remaining = numbers[:i]
            for par in self.possible_combos(remaining, target, partial + [n], partial_sum + n):
                yield par
    
    def make_combos(self):
        output = []
        for dice in range(1, 13):
            combs = []
            for res in self.possible_combos(self.arr, dice):
                combs.append(res)
            output.append(combs)
        return output

    def game_play(self, combos):
        game_going = True
        round_score = [sum(self.arr)]

        while (sum(self.arr) > 0) and game_going:
            dice = self.roll_dice()

            #moves = self.possible_combos(self.arr, dice)
            moves = combos[dice-1]
            
            combo_looking = False
            n = 0
            while not(combo_looking) and (n < len(moves)):
                drops = moves[n]
              
                if self.possible_drop(drops):
                    round_score.append(self.drop_tiles(drops))
                    combo_looking = True
                n += 1
            game_going = combo_looking
        
        round_score = [round_score[x] if x < len(round_score) else round_score[-1] for x in range(self.max_tile)]
        return round_score


if __name__ == "__main__":
    starttime = dt.datetime.now()

    arr = []
    n = 100000
    combinations = play_game().make_combos()
    for i in range(n):
        arr.append(play_game().game_play(combinations))
    arr = np.array(arr)

    final_score = arr[:, -1]

    x = np.linspace(0, 78, 79, dtype=int)
    y = np.empty_like(x)
    for i in x:
        y[i] = np.count_nonzero(final_score == i)

    y = y / n
    print(dt.datetime.now() - starttime)
    plt.plot(x, y)
    plt.show()
    print(y[0])
    
